static/app/views/dashboards/dashboardLegendUtils.tsx

```diff
@@ -0,0 +1,204 @@
+          })
+        : router.replace({
+            query: {
+              ...location.query,
+              unselectedSeries: newLegendQuery,
+            },
+          });
+    } else {
+      router.replace({
+        query: {
+          ...location.query,
+          unselectedSeries: [
+            location.query.unselectedSeries,
+            this.encodeLegendQueryParam(widget, selected),
+          ],
+        },
+      });
+    }
+  }
+
+  // sets unselected legend options by the legend query param
+  getLegendUnselected(widget: Widget) {
+    const location = this.location;
+
+    return location.query.unselectedSeries
+      ? this.decodeLegendQueryParam(widget)
+      : this.widgetRequiresLegendUnselection(widget) &&
+          this.organization.features.includes('dashboards-releases-on-charts')
+        ? {[this.encodeSeriesNameForLegend('Releases', widget.id)]: false}
+        : {};
+  }
+
+  widgetRequiresLegendUnselection(widget: Widget) {
+    return (
+      widget.displayType === DisplayType.AREA || widget.displayType === DisplayType.LINE
+    );
+  }
+
+  formatLegendDefaultQuery(widgetId?: string) {
+    return this.organization.features.includes('dashboards-releases-on-charts')
+      ? `${widgetId}-Releases`
+      : `${widgetId}-`;
+  }
+
+  // going from selected to query param
+  encodeLegendQueryParam(widget: Widget, selected: Record<string, boolean>) {
+    return (
+      widget.id +
+      '-' +
+      Object.keys(selected)
+        .filter(key => !selected[key])
+        .map(series => this.decodeSeriesNameForLegend(series))
+        .join('-')
+    );
+  }
+
+  // going from query param to selected
+  decodeLegendQueryParam(widget: Widget) {
+    const location = this.location;
+    return decodeList(location.query.unselectedSeries).reduce((acc, legend) => {
+      const legendValues = legend.split('-');
+      const widgetId = legendValues[0];
+      const seriesNames = legendValues.splice(1);
+      if (widget.id === widgetId && seriesNames) {
+        seriesNames.forEach(series => {
+          if (series) {
+            acc[this.encodeSeriesNameForLegend(series, widget.id)] = false;
+          }
+        });
+      }
+      return acc;
+    }, {});
+  }
+
+  encodeSeriesNameForLegend(seriesName: string, widgetId?: string) {
+    return `${seriesName}:${widgetId}`;
+  }
+
+  decodeSeriesNameForLegend(encodedSeriesName: string) {
+    return encodedSeriesName.split(':')[0];
+  }
+
+  // change timeseries names to SeriesName:widgetID
+  modifyTimeseriesNames(widget: Widget, timeseriesResults?: Series[]) {
+    return timeseriesResults
+      ? timeseriesResults.map(series => {
+          return {
+            ...series,
+            seriesName: this.encodeSeriesNameForLegend(series.seriesName, widget.id),
+          };
+        })
+      : [];
+  }
+
+  // when a widget has been changed/added/deleted update legend to incorporate that
+  updatedLegendQueryOnWidgetChange(newDashboard: DashboardDetails) {
```

Comment:
🤔 this one is a bit odd. Is it possible to call `encodeLegendQueryParam` for every new widget instead of this?

Reply:
for new widgets probably but it wouldn't account for if a widget was updated or deleted. That was my main reasoning for having this function, however i'm open to other suggestions (side note I just realized this function was not doing what I wanted it to do so I'm refactoring it anyways)

Reply:
🤔 I was wondering if you could do `newDashboard.widgets.forEach(widget => setWidgetSelectionState(widget, state)`. As in, use the existing state setting function on every widget in the dashboard, rather than have a new method that does this